src/Grpc/HttpApi/src/Microsoft.AspNetCore.Grpc.HttpApi/Internal/HttpApiServerCallContext.cs

```diff
@@ -0,0 +1,236 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net.Sockets;
+using System.Text;
+using System.Text.Json;
+using Grpc.AspNetCore.Server;
+using Grpc.Core;
+using Grpc.Shared;
+using Grpc.Shared.Server;
+using Microsoft.AspNetCore.Grpc.HttpApi.Internal.CallHandlers;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.AspNetCore.Mvc.Formatters;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Grpc.HttpApi.Internal;
+
+internal sealed class HttpApiServerCallContext : ServerCallContext, IServerCallContextFeature
+{
+    // TODO(JamesNK): Remove nullable override after Grpc.Core.Api update
+    private static readonly AuthContext UnauthenticatedContext = new AuthContext(null!, new Dictionary<string, List<AuthProperty>>());
+
+    private readonly IMethod _method;
+
+    public HttpContext HttpContext { get; }
+    public MethodOptions Options { get; }
+    public CallHandlerDescriptorInfo DescriptorInfo { get; }
+    public bool IsJsonRequestContent { get; private set; }
+    // Default request encoding to UTF8 so an encoding is available
+    // if the request sends an invalid/unsupported encoding.
+    public Encoding RequestEncoding { get; private set; } = Encoding.UTF8;
+
+    internal ILogger Logger { get; }
+
+    private string? _peer;
+    private Metadata? _requestHeaders;
+    private AuthContext? _authContext;
+
+    public HttpApiServerCallContext(HttpContext httpContext, MethodOptions options, IMethod method, CallHandlerDescriptorInfo descriptorInfo, ILogger logger)
+    {
+        HttpContext = httpContext;
+        Options = options;
+        _method = method;
+        DescriptorInfo = descriptorInfo;
+        Logger = logger;
+    }
+
+    public void Initialize()
+    {
+        IsJsonRequestContent = JsonRequestHelpers.HasJsonContentType(HttpContext.Request, out var charset);
+        RequestEncoding = JsonRequestHelpers.GetEncodingFromCharset(charset) ?? Encoding.UTF8;
+
+        // Add the HttpContext to UserState so GetHttpContext() continues to work
+        HttpContext.Items["__HttpContext"] = HttpContext;
```

Comment:
??

Reply:
https://github.com/grpc/grpc-dotnet/blob/7ef184f3c4cd62fbc3cde55e4bb3e16b58258ca1/src/Grpc.AspNetCore.Server/ServerCallContextExtensions.cs#L53-L61

Reply:
I can't say that made it clearer...ServerCallContext.UserState = .Items?

Reply:
https://github.com/dotnet/aspnetcore/pull/40242#pullrequestreview-908071822